packages/navie/src/services/google-vertexai-completion-service.ts

```diff
@@ -0,0 +1,137 @@
+      maxOutputTokens: 8192,
+      ...options,
+    });
+  }
+
+  get miniModelName(): string {
+    const miniModel = process.env.APPMAP_NAVIE_MINI_MODEL;
+    return miniModel ?? 'gemini-1.5-flash-002';
+  }
+
+  // Request a JSON object with a given JSON schema.
+  async json<Schema extends z.ZodType>(
+    messages: Message[],
+    schema: Schema,
+    options?: CompleteOptions
+  ): Promise<z.infer<Schema> | undefined> {
+    const model = this.buildModel({
+      ...options,
+      streaming: false,
+      responseMimeType: 'application/json',
+    });
+    const sentMessages = mergeSystemMessages([
+      ...messages,
+      {
+        role: 'system',
+        content: `Use the following JSON schema for your response:\n\n${JSON.stringify(
+          zodResponseFormat(schema, 'requestedObject').json_schema.schema,
+          null,
+          2
+        )}`,
+      },
+    ]);
+
+    for (const message of sentMessages) this.trajectory.logSentMessage(message);
+
+    const response = await model.invoke(sentMessages.map(convertToMessage));
+
+    this.trajectory.logReceivedMessage({
+      role: 'assistant',
+      content: JSON.stringify(response),
+    });
+
+    const sanitizedContent = response.content.toString().replace(/^`{3,}[^\s]*?$/gm, '');
+    const parsed = JSON.parse(sanitizedContent) as unknown;
+    schema.parse(parsed);
+    return parsed;
+  }
+
+  async *complete(messages: readonly Message[], options?: { temperature?: number }): Completion {
+    const usage = new Usage();
+    const model = this.buildModel(options);
+    const sentMessages: Message[] = mergeSystemMessages(messages);
+    const tokens = new Array<string>();
+    for (const message of sentMessages) this.trajectory.logSentMessage(message);
+
+    const maxAttempts = CompletionRetries;
+    for (let attempt = 0; attempt < maxAttempts; attempt += 1) {
+      try {
+        // eslint-disable-next-line no-await-in-loop
+        const response = await model.stream(sentMessages.map(convertToMessage));
+
+        // eslint-disable-next-line @typescript-eslint/naming-convention, no-await-in-loop
+        for await (const { content, usage_metadata } of response) {
+          yield content.toString();
+          tokens.push(content.toString());
+          if (usage_metadata) {
+            usage.promptTokens += usage_metadata.input_tokens;
+            usage.completionTokens += usage_metadata.output_tokens;
+          }
+        }
+
+        this.trajectory.logReceivedMessage({
+          role: 'assistant',
+          content: tokens.join(''),
+        });
+
+        break;
+      } catch (cause) {
+        if (attempt < maxAttempts - 1 && tokens.length === 0) {
+          const nextAttempt = CompletionRetryDelay * 2 ** attempt;
+          warn(`Received ${JSON.stringify(cause)}, retrying in ${nextAttempt}ms`);
+          await new Promise<void>((resolve) => {
+            setTimeout(resolve, nextAttempt);
+          });
+          continue;
+        }
```

Comment:
Note that this doesn't attempt to prune input tokens

Reply:
No, I figured it's unlikely to be necessary with 2M token input limit...